From 7a07c75ae48b6086dd6dc1b5f97672f9ac6228ed Mon Sep 17 00:00:00 2001
From: Daniel DeGrasse <ddegrasse@tenstorrent.com>
Date: Tue, 16 Sep 2025 11:13:40 -0500
Subject: [PATCH] arch: arc: core: support multilevel interrupt priority
 setting

Properly support setting priority for multilevel interrupts on ARC
cores.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>
---
 arch/arc/core/irq_manage.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/arch/arc/core/irq_manage.c b/arch/arc/core/irq_manage.c
index f6573f5e1bd..cc3cbf0cb7b 100644
--- a/arch/arc/core/irq_manage.c
+++ b/arch/arc/core/irq_manage.c
@@ -215,7 +215,23 @@ int arch_irq_is_enabled(unsigned int irq)

 void z_irq_priority_set(unsigned int irq, unsigned int prio, uint32_t flags)
 {
-	ARG_UNUSED(flags);
+#ifdef CONFIG_MULTI_LEVEL_INTERRUPTS
+	unsigned int level = irq_get_level(irq);
+	const struct device *dev;
+
+#ifdef CONFIG_3RD_LEVEL_INTERRUPTS
+	if (level == 3) {
+		dev = z_get_sw_isr_device_from_irq(irq);
+		irq_set_priority_next_level(dev, irq_from_level_3(irq), prio, flags);
+		return;
+} else
+#endif
+	if (level == 2) {
+		dev = z_get_sw_isr_device_from_irq(irq);
+		irq_set_priority_next_level(dev, irq_from_level_2(irq), prio, flags);
+		return;
+	}
+#endif

 	__ASSERT(prio < CONFIG_NUM_IRQ_PRIO_LEVELS,
 		 "invalid priority %d for irq %d", prio, irq);
--
2.34.1
